pkg/util/slices.go

```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
 	}
 	return a
 }
+
+// Filter returns a new slice that only contains elements from collection that
+// satisfy predicate.
+//
+//	// Filter in place
+//	numbers = Filter(numbers, isEven)
+//	// Filter into a new slice
+//	odds := Filter(numbers, isEven)
+func Filter[T any](collection []T, predicate func(T) bool) []T {
+	i := 0
+	out := make([]T, len(collection))
```

Comment:
This allocates as much as the original slice, which can be wasteful if the `predicate` filters out most of the elements.

The alternative would be to create a nil slice, and let it grow/reallocate while we add things to it. This can be considered wasteful in other cases when `predicate` doesn't filter out much.

The "universal" strategy here is taking the `out` slice as a parameter:

```golang
func Filter[T any](collection []T, predicate func(T) bool, out []) []T {
	// if predicate(item) {
	// 	out = append(out, item)
	// }
}

// Easy, but appends to the nil slice, so may allocate multiple times.
even := Filter(numbers, isEven, nil)
// Zero allocations, reuses the original slice.
numbers = Filter(numbers, isEven, numbers[:0])
// Exactly one allocation if we know/guessed the number of elements 
even := Filter(numbers, isEven, make([]int, 0, 10))
```